### PR TITLE
Refactor weather code plugins and CLIs

### PR DESCRIPTION
**Pull request by MoseleyS**
Time spent on this ticket should be charged to a specific code. Ask @MoseleyS.

As Enhancing Post Processing Science Lead, I want the IMPROVER weather code plugin to be refactored so that I can run a precipitation type code configuration.

The essential changes are around metadata. The output cube name and categorical attributes are now derived from the decision tree, which has been updated as a result and is now also a required input to the modal categories plugin.

To achieve this, I have done some refactoring:

- code and acceptance test data moved to directories named `categorical`
- CLI names changed:
  - `wxcode` -> `categorical`
  - `wxcode-modal` -> `categorical-modes`
- Both these CLIs now require the decision tree, with the key name changing from `wxtree` to `decision-tree`
- Plugin names changed:
  - `WeatherSymbols` -> `ApplyDecisionTree`
  - `ModalWeatherCode` -> `ModalCategory`
- Several internal variable names and comments have been updated to be more generic.
- Meta data definitions of the cube name, attribute names and contents have moved out of the code and into the decision tree.
- Grouping and day -> night mapping have moved out of the code and into the decision tree.

Acceptance test data updates in https://github.com/metoppv/improver_test_data/pull/28

For reviewing, I recommend looking in two parts:

1. Most of the refactoring is in the first four commits: aaf3ae6c...5734e30
2. Then everything else: 5734e30...99f68020

This PR **MUST** be merged with the accompanying suite PR: (https://github.com/MetOffice/improver_suite/pull/1823)

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Review by mspelman07 — CHANGES_REQUESTED** [review]
Thanks Stephen. A few comments mostly around clarifying some docstrings but otherwise this looks good.

File: doc/source/extended_documentation/categorical/build_a_decision_tree.rst [hunk 1]
```diff
@@ -26,7 +27,19 @@ forecast, proceed to the if_true node, else move to the if_false node.
...
 **Encoding a decision tree**
 
-The first node above is encoded as follows::
+The meta node provides the name to use for the metadata of the resulting cube and
+can be anywhere in the decision tree.
+This becomes the cube name and is also used for two attributes that describe the
+categorical data: **<name>** and **<name>_meaning**::
+
+  {
+    "meta": {
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply doc/source/extended_documentation/categorical/build_a_decision_tree.rst:36]: Clause added

File: improver/categorical/decision_tree.py [hunk 2]
```diff
@@ -158,7 +162,21 @@ def __init__(
 
     def __repr__(self) -> str:
         """Represent the configured plugin instance as a string."""
-        return "<WeatherSymbols start_node={}>".format(self.start_node)
+        return "<ApplyDecisionTree start_node={}>".format(self.start_node)
+
+    @staticmethod
+    def _is_decision_node(key: str, query: Dict) -> bool:
+        """
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:169]: Added

File: improver/categorical/decision_tree.py [hunk 3]
```diff
@@ -485,9 +501,9 @@ def remove_optional_missing(self, optional_node_data_missing: List[str]):
             alternative = self.queries[missing][target]
 
             for node, query in self.queries.items():
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:503]: Changed

File: improver/categorical/decision_tree.py [hunk 4]
```diff
@@ -532,22 +548,22 @@ def find_all_routes(
...
                     graph, node, end, route=route
                 )
                 routes.extend(newroutes)
         return routes
 
-    def create_symbol_cube(self, cubes: Union[List[Cube], CubeList]) -> Cube:
+    def create_categorical_cube(self, cubes: Union[List[Cube], CubeList]) -> Cube:
         """
-        Create an empty weather symbol cube
+        Create an empty categorical cube
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:559]: Updated

**Review by mspelman07 — CHANGES_REQUESTED** [review]
I'm happy with the changes made. 

The only remaining references to weather symbols that I can find (ignoring tests) are in the following locations:

improver.categorical.utilities.categorical_attributes
improver.developer_tools.metadata_interpreter.MOMetadataInterpreter (in function run)

The categorical utilities just has a few variables called wx_keys which probably wants updating but I'm not sure about the developer_tools one?

**Comment by MoseleyS** [comment]
> improver.categorical.utilities.categorical_attributes improver.developer_tools.metadata_interpreter.MOMetadataInterpreter (in function run)
> 
> The categorical utilities just has a few variables called wx_keys which probably wants updating but I'm not sure about the developer_tools one?

I've updated the variable names in both files. This allows us to add other categorical data to the metadata interpreter in the future, if required.

**Review by mspelman07 — DISMISSED** [review]

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks @MoseleyS 👍 

I've added some very minor comments.

File: doc/source/extended_documentation/categorical/build_a_decision_tree.rst [hunk 5]
```diff
@@ -93,8 +102,35 @@ accessed with this key contains the essentials that make the node function.
     against the spp__relative_to_threshold attribute of the threshold coordinate
     in the provided diagnostic.
 
+The first leaf node above is encoded as follows::
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply doc/source/extended_documentation/categorical/build_a_decision_tree.rst:105]: Updated

File: improver/categorical/decision_tree.py [hunk 6]
```diff
@@ -158,7 +162,25 @@ def __init__(
 
     def __repr__(self) -> str:
         """Represent the configured plugin instance as a string."""
-        return "<WeatherSymbols start_node={}>".format(self.start_node)
+        return "<ApplyDecisionTree start_node={}>".format(self.start_node)
+
+    @staticmethod
+    def _is_decision_node(key: str, query: Dict) -> bool:
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:168]: Thanks. I don't think I could work out a complete type hint that was correct.

File: improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py [hunk 7]
```diff
@@ -41,13 +41,12 @@
 from iris.exceptions import CoordinateNotFoundError
 from iris.tests import IrisTest
 
+from improver.categorical.decision_tree import ApplyDecisionTree
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/categorical/decision_tree/test_ApplyDecisionTree.py:44]: Changed

File: improver/categorical/decision_tree.py [hunk 8]
```diff
@@ -559,7 +584,7 @@ def create_symbol_cube(self, cubes: Union[List[Cube], CubeList]) -> Cube:
         mandatory_attributes = generate_mandatory_attributes(cubes)
         if self.title:
             mandatory_attributes.update({"title": self.title})
-        optional_attributes = weather_code_attributes()
+        optional_attributes = categorical_attributes(self.queries, self.meta["name"])
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:587]: Changed

File: improver/categorical/decision_tree.py [hunk 9]
```diff
@@ -800,49 +824,46 @@ def process(self, cubes: CubeList) -> Cube:
...
                         ) = self.invert_condition(current)
-
-                    conditions.append(self.create_condition_chain(current))
+                    if "leaf" not in current.keys():
+                        conditions.append(self.create_condition_chain(current))
                 test_chain = [conditions, "AND"]
 
                 # Set grid locations to suitable weather symbol
-                symbols.data[
+                categories.data[
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:863]: Done

File: improver/categorical/modal_code.py [hunk 10]
```diff
@@ -48,106 +48,124 @@
...
-class ModalWeatherCode(BasePlugin):
-    """Plugin that returns the modal code over the period spanned by the
+class ModalCategory(BasePlugin):
+    """Plugin that returns the modal category over the period spanned by the
     input data. In cases of a tie in the mode values, scipy returns the smaller
     value. The opposite is desirable in this case as the significance /
-    importance of the weather codes generally increases with the value. To
-    achieve this the codes are subtracted from an arbitrarily larger
+    importance of the weather code categories generally increases with the value. To
+    achieve this the categories are subtracted from an arbitrarily larger
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/modal_code.py:59]: I've modified this, but not quite to your suggestion.

File: improver/categorical/decision_tree.py [hunk 11]
```diff
@@ -570,7 +595,7 @@ def create_symbol_cube(self, cubes: Union[List[Cube], CubeList]) -> Cube:
             )
 
         symbols = create_new_diagnostic_cube(
-            "weather_code",
+            self.meta["name"],
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py:598]: changed

File: improver/categorical/utilities.py [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+            items.extend([item])
+    return items
+
+
+def update_daynight(cube: Cube, day_night: Dict) -> Cube:
+    """ Update category depending on whether it is day or night
+
+    Args:
+        cube:
+            Cube containing only daytime weather symbols.
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/utilities.py:172]: I'm surprised at how many of these I missed! Changed.

File: improver/categorical/utilities.py [hunk 13]
```diff
@@ -0,0 +1,565 @@
...
+            # Check that this leaf is reachable, or is declared unreachable.
+            if not ((items.get("is_unreachable", False)) or node in all_targets):
+                issues.append(
+                    f"Unreachable leaf '{node}'. Add 'is_unreachable': True to suppress this issue."
+                )
+            if (items.get("is_unreachable", False)) and node in all_targets:
+                issues.append(f"Leaf '{node}' has 'is_unreachable' but can be reached.")
+
+            # If leaf key is present, check it is an int.
+            leaf_target = items.get("leaf", -1)
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/utilities.py:417]: Yes. Changed.

File: improver/categorical/utilities.py [hunk 14]
```diff
@@ -0,0 +1,565 @@
...
+                    f"and diagnostic_fields: {prob_thresholds}, {diagnostic_fields}"
+                )
+
+    if not issues:
+        issues.append("Decision tree OK\nRequired inputs are:")
+        issues.append(interrogate_decision_tree(decision_tree))
+    return "\n".join(issues)
+
+
+def day_night_map(decision_tree) -> Dict:
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/utilities.py:551]: I've added your earlier suggestion here too.

File: improver_tests/categorical/decision_tree/__init__.py [hunk 15]
```diff
@@ -35,8 +35,8 @@
 
 import numpy as np
 
+from improver.categorical.utilities import categorical_attributes
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/categorical/decision_tree/__init__.py:38]: Done

File: improver_tests/categorical/test_utilities.py [hunk 16]
```diff
@@ -152,8 +115,43 @@ def setUp(self):
         self.cube = add_coordinate(
             cube, date_times, "time", is_datetime=True, order=[1, 0, 2, 3],
         )
-        self.wxcode = np.array(list(WX_DICT.keys()))
-        self.wxmeaning = " ".join(WX_DICT.values())
+        self.decision_tree = wxcode_decision_tree()
+        self.wxcode = np.arange(31)
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver_tests/categorical/test_utilities.py:119]: Done something equivalent.

File: improver/categorical/decision_tree.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/decision_tree.py]: Agreed. I've moved it into utilities and added tests. It is now used in update_tree_thresholds too.

**Review by gavinevans — CHANGES_REQUESTED** [review]
Thanks for the updates, @MoseleyS 👍 

I've added one trivial comment.

File: improver/categorical/modal_code.py [hunk 17]
```diff
@@ -48,106 +48,124 @@
...
+    """Plugin that returns the modal category over the period spanned by the
     input data. In cases of a tie in the mode values, scipy returns the smaller
     value. The opposite is desirable in this case as the significance /
-    importance of the weather codes generally increases with the value. To
-    achieve this the codes are subtracted from an arbitrarily larger
-    number prior to calculating the mode, and this operation reversed in the
-    final output.
+    importance of the weather code categories generally increases with the value. To
+    achieve this the categories are subtracted from an arbitrarily larger
+    number prior to calculating the mode, and this operation reversed before the
```
[the remark this answers is not in the document]
Comment (MoseleyS) [reply improver/categorical/modal_code.py]: Grammar corrected

**Review by gavinevans — APPROVED** [review]

**Pull request merged**